### DISCUSSION TITLE
Is there a way to switch to recent os window?

### DISCUSSION THREAD

**Discussion by jackielii** [kovidgoyal/kitty#7007] [Q&A]
I'm trying to setup my kitty to fully replace Tmux, one thing I use very often is tmux session switching. Specifically switch to last session.

In kitty, there is no session, but I use a separate OS window for a session which works great. But there doesn't seem to be a way to switch to the "recent focused" OS window.

There is the `nth_os_window` mappable action, but it doesn't accept `-1` to switch to last-focused window like `nth_window -1` does.

In `boss.py` there also doesn't seem to be a way. Is this possible?

**Comment by kovidgoyal** [comment]
Not presently, someone would have to implement a history system for OS
windows like the one for kitty windows and then allow nth_os_window to
accept negative numbers.

**Reply by jackielii** [reply]
would you accept a PR for this?

**Comment by kovidgoyal** [comment]
yes, sure. In fact, it should be pretty trivial,
use the existing last_focused_counter (you will need to create a new integration
point with python to access it) to implement nth_os_window(-n)

**Reply by jackielii** [reply]
If I understood correctly, we're looking for the window with second largest `os_window->last_focused_counter` to get the previous focused window right?

https://github.com/kovidgoyal/kitty/blob/master/kitty/state.c#L107-L117

I'm not sure I understand this function. It seems to always find the largest last_focused_counter?

**Reply by jackielii** [reply]
I submitted #7009 on this assumption.

**Comment by kovidgoyal** [comment]
On Wed, Jan 17, 2024 at 07:37:18AM -0800, Jackie Li wrote:
> If I understood correctly, we're looking for the window with second largest `os_window->last_focused_counter` to get the previous focused window right?
> 
> https://github.com/kovidgoyal/kitty/blob/master/kitty/state.c#L107-L117
> 
> I'm not sure I understand this function. It seems to always find the largest last_focused_counter?

It's for the case when no kitty OS window is focused, so it finds the
window that last had focus.
